src-electron/db/query-loader.js

```diff
@@ -1151,35 +1151,81 @@ async function getEndpointCompositionIdByCode(db, deviceType) {
- * DEVICE_COMPOSITION table. It handles the insertion of the device code,
- * endpoint composition reference, conformance, and constraint values.
+ * DEVICE_COMPOSITION table for each deviceType in each endpoint. It handles the insertion
+ * of the device code, endpoint composition reference, conformance, and constraint values.
  * Note that the "CONSTRAINT" column name is escaped with double quotes
  * to avoid conflicts with the SQL reserved keyword.
  *
  * @param {Object} db - The database connection object.
  * @param {Object} deviceType - The device type object containing the data to be inserted.
  * @param {number} endpointCompositionId - The ID of the endpoint composition.
- * @returns {Promise} A promise that resolves when the insertion is complete.
+ * @returns {Promise} A promise that resolves when all insertions are complete.
  */
 function insertDeviceComposition(db, deviceType, endpointCompositionId) {
-  const insertQuery = `
-    INSERT INTO DEVICE_COMPOSITION (CODE, ENDPOINT_COMPOSITION_REF, CONFORMANCE, DEVICE_CONSTRAINT)
-    VALUES (?, ?, ?, ?)
-  `
-  try {
-    return dbApi.dbInsert(db, insertQuery, [
-      parseInt(deviceType.childDeviceId, 16),
-      endpointCompositionId,
-      deviceType.conformance,
+  // Ensure that deviceType and its necessary properties are defined
+  if (
+    !deviceType ||
+    !deviceType.composition ||
+    !deviceType.composition.endpoint
+  ) {
+    throw new Error('Invalid deviceType object or endpoint data')
+  }
+
+  // Make sure 'deviceType.composition.endpoint' is always an array, even if there's only one endpoint
+  const endpoints = Array.isArray(deviceType.composition.endpoint)
+    ? deviceType.composition.endpoint
+    : [deviceType.composition.endpoint]
+
+  // Prepare an array to hold all insert queries
+  const insertQueries = []
+
+  // Iterate over all endpoints in the deviceType and their respective deviceTypes
+  for (let i = 0; i < endpoints.length; i++) {
+    const endpoint = endpoints[i]
+
+    // Ensure deviceType is present and handle both single value or array
+    const deviceTypes = Array.isArray(endpoint.deviceType)
+      ? endpoint.deviceType
+      : endpoint.deviceType
+        ? [endpoint.deviceType]
+        : [] // Default to empty array if undefined
+
+    // Use the $ to get the endpoint-specific conformance and constraint values
+    const endpointConformance =
+      endpoint.endpointComposition?.endpoint?.$.conformance ||
+      deviceType.conformance
+    const endpointConstraint =
+      endpoint.endpointComposition?.endpoint?.$.constraint ||
       deviceType.constraint
-    ])
-  } catch (error) {
-    console.error('Error inserting device composition:', error)
-    throw error // Re-throw the error after logging it
+
+    // Create insert queries for each deviceType in this endpoint and add them to the insertQueries array
+    for (let j = 0; j < deviceTypes.length; j++) {
+      const device = deviceTypes[j]
+
```

Comment:
Why not:
```
   for ( let device of deviceTypes) { 
      ...
   }
```
I don't think you're using the actual index anywhere, and while there is something cozy about the good old fashioned 1960's C-like for loop, it is visually more taxing than modern JS.